ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEventListener.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.query.QueryError;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.QueryMetadata;
+import java.util.Optional;
+import org.apache.kafka.streams.KafkaStreams;
+
+/**
+ * Defines the events that clients of {@link io.confluent.ksql.query.QueryRegistry} can register
+ * to listen on.
+ */
+public interface QueryEventListener {
```

Comment:
I went back and forth between what's implemented here, and a structure with 1 interface (e.g. `QueryRegistrationListener`), which listens on new/deregister queries, and another interface like this one for events on a single query. In this model you'd add the listeners after the query is created (like we do before this patch). Something like:

```
interface QueryRegistrationListener {
  void onCreate(...) {}
  void onDeregister(...) {}  
  void onSandbox(QueryRegistry sanboxRegistry) {}
}

interface QueryRegistry {
    void addQueryRegistrationListener(QueryRegistrationListener l);
}

interface QueryEventListener {
    void onStateChange(...){}
    void onError(...){}
    void onClose(...){}
    void onSandbox(QueryMetadata sandboxQueryMetadata) {}
}

class QueryMetadata {
    void addQueryEventListener(QueryEventListener l);
}
```

Then you would write code like:
```
class SomethingThatCaresAboutQueries implements QueryRegistrationListener {
    void onCreate(...) {
        query.addListener(new PerQueryListener());
    }

    void onSandbox(QueryRegistry sandbox) {
        for (QueryMetadata sandboxQuery : sandbox.getAllLiveQueries()) {
            sandboxQuery.addListener(/* create a listener for the sandbox to react to sandbox events */);
        }
    }

    class PerQueryListener implements QueryEventListener {
         // per-query state and code that listens and reacts to events
    }
}
```

They're functionally the same, but keeping it all in one interface seemed a little cleaner and easier to follow, at the expense of needing to track a mapping from query id to per-query-state in the listener classes.

Reply:
In my view, `QueryMetadata` is meant to model a single query, so in that case we'd still have a different `QueryMetadata` for each query.  We'd need to split out some of this functionality into another class that models a Kafka Streams runtime, and tracks the queries assigned to each runtime. The state change and error notifications should go into that class, and QueryStateListener should listen on that instead of QueryMetadata as it does today. So, as a client of the engine you can register to be notified on per-query events, or per-runtime events depending on the use case. Some use cases for the former would be per-query cleanup on terminate, per-query limits checking (e.g. whats in https://github.com/confluentinc/ksql/pull/7323), and tracking errors for each query that can be surfaced in the API. A use case for the latter would be emitting metrics on the current status and error state of each runtime.